### PR TITLE
[205] New <fileMappers> option

### LINKED ISSUE #205: FileMappers
Sometimes it is needed to not unpack *all* content, but just *some*. The Maven Dependency Plugin's unpack task therefore has the ability to apply `FileMappers`. It would be a great addition to the Download Maven Plugin to apply `FileMappers`, too!

### PR DESCRIPTION
**Pull request by mkarg**
Resolves #205

### CODE REVIEW

**Comment by mkarg** [comment]
@longtimeago I would be very happy if you could integrate this PR as we need this common-purpose feature very urgently to rewrite ZIPs while unpacking them.

File: src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java [hunk 1]
```diff
@@ -625,4 +635,26 @@ private boolean useHttpProxy(final ProxyInfo proxyInfo) {
...
         return result;
     }
+
+    /**
+     * @return {@link FileMapper}s to be used for rewriting each target path, or {@code null} if no rewriting shall
+     *         happen.
+     *
+     * @since 1.6.8
+     */
+    public FileMapper[] getFileMappers()
```
Comment (longtimeago) [inline src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java:645]: @mkarg do we really need getter/setter for `fileMappers`?

File: src/it/GetAndUnpackWithFileMapper/pom.xml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+                        <phase>generate-resources</phase>
+                        <goals>
+                            <goal>wget</goal>
+                        </goals>
+                        <configuration>
+                            <url>http://downloads.sourceforge.net/ant-contrib/ant-contrib-1.0b2-bin.zip</url>
+                            <unpack>true</unpack>
+                            <fileMappers>
+                                <org.codehaus.plexus.components.io.filemappers.RegExpFileMapper>
+                                    <pattern>README.txt</pattern>
```
Comment (longtimeago) [inline src/it/GetAndUnpackWithFileMapper/pom.xml:29]: @mkarg technically, this pattern will match any of these names
- `README.txt`
- `README_txt`
- `READMEtxt`
..

I'd escape the dot to be sure it matches the dot only and not any char.
Mostly for readability ;)

**Review by longtimeago — COMMENTED** [review]
@mkarg very nice idea!

Just 2 minor suggestions :)

File: src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java (hunk 1 above)
Comment (mkarg) [reply src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java:645]: I don't know actually if the solution still works without. I simply copied from the dependency plugin source code.

Comment (mkarg) [reply src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java:645]: If you want me to remove it, please clearly tell me.

File: src/it/GetAndUnpackWithFileMapper/pom.xml (hunk 2 above)
Comment (mkarg) [reply src/it/GetAndUnpackWithFileMapper/pom.xml:29]: I can change that, but for the actualy outcome this makes no difference, as the `README.txt` file actually is the sole one matching that pattern.

File: src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java (hunk 1 above)
Comment (longtimeago) [reply src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java:645]: It should work, so, please, remove )

File: src/it/GetAndUnpackWithFileMapper/pom.xml (hunk 2 above)
Comment (longtimeago) [reply src/it/GetAndUnpackWithFileMapper/pom.xml:29]: Sure, it's just for self documentation purpose )

File: src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java (hunk 1 above)
Comment (mkarg) [reply src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java]: Just tested and confirm that it passes test without, so I removed it in https://github.com/maven-download-plugin/maven-download-plugin/pull/206/commits/d7b02e0f5d5ce439a9e1c54e8c7c842803606434.

File: src/it/GetAndUnpackWithFileMapper/pom.xml (hunk 2 above)
Comment (mkarg) [reply src/it/GetAndUnpackWithFileMapper/pom.xml]: I can't see neiter benefit nor drawback (in fact it currently documents very well that wildcards can be used), so just did it in https://github.com/maven-download-plugin/maven-download-plugin/pull/206/commits/01c2c2c91a5e0548804e19b3ebc5c4b2283ede75 to fulfil your personal taste. ;-)

Comment (longtimeago) [reply src/it/GetAndUnpackWithFileMapper/pom.xml]: @mkarg thank you, sir! :+1:

**Review by longtimeago — APPROVED** [review]
@mkarg thank you very much! :+1:

**Comment by mkarg** [comment]
@longtimeago Thanks for adopting my contribution. :-)

**Pull request merged**